Fixing Bug#7554562 - Supporting "!include" and "!includedir" directives

kapilash opened the pull request:
This commit introduces support for "include" and "includedir" directives in MySQL configuration files

kapilash commented:
cc @jeffaco @terlochan

jeffaco commented on build/Makefile:
```diff
@@ -384,6 +384,7 @@ STATIC_PROVIDERLIB_SRCFILES += \
 	$(PROVIDER_DIR)/support/mycnf-paths.cpp \
 	$(PROVIDER_DIR)/support/productdependencies.cpp \
 	$(PROVIDER_DIR)/support/sqlauth.cpp \
+	$(PROVIDER_DIR)/support/mycnf-scanner.cpp \
```
Do you actually need this source file in the provider library? Please verify this.

Given that we can't necessarily read the MySQL configuration files when we're not privileged, I wouldn't expect that you'd need this from the provider itself (the provider runs without privileges). Please investigate, thanks.

kapilash replied:
What we are introducing now seemed some sort of mysql specific "support" functionality - like mycnf-paths.cpp and sqlauth.cpp. so I moved the files here.

Moreover, for some reason that I do not quite understand, [sqlbinding.cpp](https://github.com/Microsoft/MySQL-Provider/blob/master/source/code/providers/support/sqlbinding.cpp#L136) in this folder is reading the cnf files. That line gets executed  at least by the myauth executable (which also runs with root permissions, btw). At some point in future, sqlbinding.cpp may need a subset of the scanner functionality that we are introducing, because mysql_options does not support "!include" and "!includedir"

If you think my above concerns are invalid, I will move this to [omi_prexec](https://github.com/Microsoft/MySQL-Provider/tree/master/source/code/omi_preexec) folder

jeffaco replied:
The problem is that the provider can't reliably to this. It is platform specific, but on some platforms, the MySQL configuration files are not W:R. Since the provider runs as-requestor, that means that a non-root user would not be able to see MySQL stats, and that wasn't acceptable to Kris Bash, our PM.

As a result, the MySQL configuration files are read exclusively by omi_preexec, which runs with privileges. That ends up setting up the auth file, allowing the provider to run without root privileges. If sqlbinding.cpp is reading the config files, that's an error, I believe, and one that you need to track down. Remember, this code used to live exclusively in omi_preexec, so sqlbinding.cpp reading the config files wasn't previously possible (I believe).

Since the provider itself should not read the cnf files (and indeed, on some platforms, CAN'T read the conf files), the provider itself should not be accessing these routines. It should be harmless to remove them.

What sqlbinding.cpp DOES do is read the auth file to get the sql ports. But that's different than reading the cnf files.

---
In reply to: [66894783](https://github.com/Microsoft/MySQL-Provider/pull/5#discussion_r66894783) [](ancestors = 66894783)

kapilash replied:
OK. Thanks for explaining. I will move it to the omi_prexec.

jeffaco commented on build/Makefile:
```diff
@@ -392,6 +393,7 @@ STATIC_PROVIDERLIB_LOGPOLICY += \
 PROVIDER_HEADERS = \
 	$(PROVIDER_DIR)/support/logstartup.h \
 	$(PROVIDER_DIR)/support/sqlauth.h \
+	$(PROVIDER_DIR)/support/mycnf-scanner.h \
```
Ditto for above.

jeffaco commented on build/Makefile:
```diff
@@ -480,7 +484,8 @@ STATIC_PROVIDER_UNITFILES = \
 	$(TEST_DIR)/myauth/myauth_test.cpp \
-	$(PROVIDER_TEST_DIR)/providertestutils.cpp
+	$(PROVIDER_TEST_DIR)/providertestutils.cpp \
+	$(PROVIDER_TEST_DIR)/config_scanner_test.cpp
```
Won't matter for build, but this should be moved higher up in the list of tests. The intent was that tests themselves are first, and support files follow. Given that, providertestutils.cpp should probably be the last one. And that said, myauth_test.cpp (which you didn't touch) should be moved up too, since testrunnerlogpolicy.cpp is a support file. Very minor nit.

jeffaco commented on source/code/omi_preexec/omi-mysql.cpp:
```diff
@@ -82,113 +83,19 @@ static int MySQL_ConfigurationParser(uid_t uid, gid_t gid, const std::wstring& c
-    catch (const SCXException& e)
-    {
-        SCX_LOGERROR(hLog, StrAppend(L"Error reading configuration file: ", e.What()));
+        SCX_LOGERROR(hLog, StrAppend(L"Error reading ", cnfFile));
```
This is a poor error. If your class can throw, then you're responsible for having meaningful exceptions, and you should at least say what the exception is (like I used to do in the old code). But this assumes that your class raises meaningful exceptions in case of errors.

Seeing an error like: "Error reading <filename>" is not helpful to a customer.

kapilash replied:
The class is not throwing exceptions - all the exceptions are already logged and indicated. This log statement is an additional log on top of an exception.
 I will get into the reason for not throwing exception from the class in the next comment.

jeffaco replied:
Did you get to this? I didn't see that ...

---
In reply to: [66894954](https://github.com/Microsoft/MySQL-Provider/pull/5#discussion_r66894954) [](ancestors = 66894954)

jeffaco commented on source/code/omi_preexec/omi-mysql.cpp:
```diff
@@ -82,113 +83,19 @@ static int MySQL_ConfigurationParser(uid_t uid, gid_t gid, const std::wstring& c
-    std::vector<std::wstring> lines;
-    try
+    MySQL::MySQL_ConfigurationScanner configScanner;
+    if (!configScanner.Scan(cnfFile, auth))
```
It looks like the Scan() method just handles all of the reading/parsing of the file, handling recursion automatically. Nice design, that's exactly what I would have done.

That also has the benefit of encapsulating all config file handling to one place, which is great.

jeffaco commented on source/code/providers/support/mycnf-scanner.h:
```diff
@@ -0,0 +1,155 @@
+#include <scxcorelib/scxlog.h>
+#include <scxcorelib/stringaid.h>
+
+#include "sqlauth.h"
```
Do you really need all of these include files in this include file? Seems like a lot - please trim this list down if possible. If you need 'em, you need 'em. But I'm wondering if you cut/pasted from elsewhere and don't actually need all of these.

kapilash replied:
I will look into these, thanks. I think one/two of ```scxfile.h```, ```scxfilepath.h``` and ```scxfilesystem.h``` are unnecessary. 
Others are needed.

jeffaco commented on source/code/providers/support/mycnf-scanner.h:
```diff
@@ -0,0 +1,155 @@
+
+       MySQL_ConfigurationScanner()
+           : m_extDepends(new MySQL_CnfScannerDependency()),
+             m_log(SCXCoreLib::SCXLogHandleFactory::GetLogHandle(L"mysql.cwconfigurationscanner"))
```
What is "cw" here, and why do you have "mysql.cwconfigurationscanner" for one constructor and "mysql.configurationscanner" for another? These should probably be the same, right?

kapilash replied:
My bad.  I will fix it. (The one time I use  vi, world will get to know of it :cry):

jeffaco commented on source/code/providers/support/mycnf-scanner.cpp:
```diff
@@ -0,0 +1,244 @@
+
+namespace MySQL
+{
+    static inline bool GetPathFromDirective(const std::wstring& line, SCXCoreLib::SCXFilePath& path)
```
No Doxygen heads? I have that in other places of MySQL, please maintain that.

Also, please have a comment of exactly what you expect the line that you're parsing to look like (i.e. give examples) so someone can quickly look at this and understand.

jeffaco commented on source/code/providers/support/mycnf-scanner.cpp:
```diff
@@ -0,0 +1,244 @@
+    static inline bool GetPathFromDirective(const std::wstring& line, SCXCoreLib::SCXFilePath& path)
+    {
+        std::vector<std::wstring> elements;
+        SCXCoreLib::StrTokenize(line, elements, L" ");
```
Are tabs allowed in the syntax? I'm guessing "yes", and that you're being too restrictive here.

jeffaco commented on source/code/providers/support/mycnf-scanner.cpp:
```diff
@@ -0,0 +1,244 @@
+        const SCXCoreLib::SCXFilePath& file, 
+        MySQL_Authentication& auth,
+        std::set< std::wstring >& visited, 
+        std::map<unsigned int, std::wstring>& ports)
```
Hmm. You have an object. Why pass the same parameters over and over again? Can't you just store these once in the object and then let the class methods manipulate these?

The filename changes all the time, I'd think, so that needs to be passed. But the others? Now that you have a class, change it to take care of it's own containing object and don't force the caller to worry about that.

kapilash replied:
I did this way to have a straight forward semantics for the ```Scan``` method. I want to keep it stateteless (ignoring the logger here).

Basically, I was trying to define the meaning of following unlikely scenario:

     MySQL_ConfigurationScanner scanner;
     scanner.Scan(file1, auth1);
     scanner.Scan(file2, auth2); // it should start with clean state - visited is empty

The above semantics are possible to achieve, even with state in the class.
But, with the state. the header file of ```MySQL_ConfigurationScanner``` will not indicate the user as to what could happen.

If it is the same either ways for you, I would like to keep it stateless in the next iteration. But if you have strong opinions, I'll  change it to a stateful class. Please let me know.

jeffaco replied:
I don't see a use-case for this. By keeping state inside of the object, you "hide" complexity (by having fewer parameters), and you allow for the Scan method to change implementations. Right now, the implementation is somewhat fixed because of what you're passing for state.

Besides, setting up logging is non-trivial (getting the m_log handle, etc). It's better to have the constructor take care of that once for the life of the object.

If there was a compelling use case, you could make this a static member and pass state. But, so far, we never do that unless there's a compelling need. I don't see a compelling need here, and I see reasons to not do that (performance, etc). In most cases, we do not do static classes unless, for some reason, we really need that. And, often, that makes maintenance of the class harder if we need to make changes later (like I just did recently to the PAL networking class).

I would strongly recommend changing the class to take care of it's own state, thus hiding implementation within the class. This is basic Object Design, and makes for more maintainable code.

---
In reply to: [66896591](https://github.com/Microsoft/MySQL-Provider/pull/5#discussion_r66896591) [](ancestors = 66896591)

kapilash replied:
I am not sure I agree with you here. But  I will change it the way you are like :+1:

jeffaco replied:
Well, I'm not sure I'm comfortable with this.

I think I'm right here for how WE designed the code. However, you're taking this over. In the end, it's you that's going to maintain this and support it going forward. Given that, you should have decision making authority for how the code is designed as long as it works properly and meets product requirements.

I feel comfortable in saying that putting implementation details in the object implementation makes perfect sense, and I view this as "object oriented design". It allows you to completely change the implementation of the object without ever changing the caller (a good thing).

But that said: You need to maintain the code. Given that, you should be comfortable with the code base and making decisions on your own.

Please discuss this with your team mates. In the end, I want you to do what you think is right.

My reviews are to catch some design errors (and there were some regarding privileges) and otherwise confirm to our coding standards. However, they are OUR coding standards (for my group), and may not apply to your group.

Please discuss with your team mates. This is a style issue and, if you really feel I'm wrong, do it as you wish. It won't affect product design.

jeffaco commented on source/code/providers/support/mycnf-scanner.cpp:
```diff
@@ -0,0 +1,244 @@
+        {
+            std::wstringstream ss;
+            ss << L"Already visited " << file.Get() << L". skipping";
+            SCX_LOGWARNING(m_log, ss.str());
```
I'm not that far yet. But it seems like this is written in a way to not allow testing this capability. I guess I'll see that when I get to the unit test.

jeffaco commented on source/code/providers/support/mycnf-scanner.cpp:
```diff
@@ -0,0 +1,244 @@
+                }
+                else
+                {
+                    SCX_LOGWARNING(m_log, SCXCoreLib::StrAppend(L" Ignoring Error in !includedir directive.", line));
```
That error leaves something to be desired. WHY was this included? Was that already logged in all cases? If so, you might want to change to something like:

Ignoring include directive "<line from file>" due to previous error in the log

Something like that, anyway.

kapilash replied:
> WHY was this included?

When the control is here, we have reached at a "wrong" line in cnf. Say user wrote something like   ```!includedir  foo bar```   

The [documentation](http://dev.mysql.com/doc/refman/5.7/en/option-files.html) does not tell us what MySQL will do in error cases like this. I think I should not throw an exception in this case. At the same time, the line is actually starting with "!includedir". So clearly a user error.
So, I am logging it.

> Ignoring include directive "" due to previous error in the log

I will try to make it more informative.

jeffaco replied:
I'm not sure what you're saying is true.  What if I have "!includedir foo bar".  Would "foo bar" be interpreted as a directory (remember, while "bad form", spaces are allowed in directory names)? If so, that would make this a legal statement.

In general, on lines that we parse, if we find something invalid, we should report it right there and then and probably stop reading. This makes it easier to find parsing code down the road for specific file sections and change parsing as necessary.

In this particular case, the parser is still pretty "dumb"; there's a lot of lines we just don't care about and ignore. But for what we do care about, if we find something we don't like, we should just log and abort.

---
In reply to: [66897123](https://github.com/Microsoft/MySQL-Provider/pull/5#discussion_r66897123) [](ancestors = 66897123)

kapilash replied:
The particular example that I gave (spaces in directory) is not acceptable for includedir or include.
I verified it.

> In general, on lines that we parse, if we find something invalid, we should report it right there and then and probably stop reading. This makes it easier to find parsing code down the road for specific file sections and change parsing as necessary.

OK. I will change it in my next iteration. 
But I thought, in general, we must be more lenient towards the cnf file than MySQL. As in, we must never treat as error what MySQL sees as valid. But the converse is okay.
That is why I eschewed all exceptions.
 I just ignore them and processed the rest that I could process. If it is an error, we will not be able to connect.

jeffaco replied:
There are errors where we'll stop anyway. If an error occurs where we know we'll do the wrong thing (like ignore an includedir or something), then we shouldn't do a "half" approach, reading some of it and not others. Better to just stop, log the issue, and leave the auth file alone. If we can't run properly, then it's better to just stop rather than to essentially "mangle" the auth file by reading some of the config but not others.

kapilash replied:
OK. I will change it the way you are like :+1:

jeffaco commented on test/code/providers/config_scanner_test.cpp:
```diff
@@ -0,0 +1,669 @@
+        std::wstring cnfFileName = L"NoDirectives.cnf";
+        std::wstring bindAddress = L"172.22.13.1";
+        unsigned int port = 3505;
+        CreateConfig(bindAddress, port, true, contents);
```
I'm trying hard to avoid a "not invented here" syndrome, and I don't think I am. But I'm not thrilled with the way that you handled this. I sort of expected something similar to the tests that GetLinuxOS has, in that we check in the real-live configuration files on the machine (and I suppose we could check in our test onesas well). It's easy to extend that, you have a bunch of "real-life" examples of actual config files that ship with various systems (like 10 or 12 of them).

Here's what I'm sort of thinking I'd like to see in the unit tests:

1. You create a directory (like test/code/...), that contains support files for various platforms. Again, see what was done for GetLinuxOS. You can check in files from numerous systems, as well as creating directories and files for specific test cases that you're interested in,

2. You get rid of CreateConfig and just read in those files (giving the path in a dependency object, and that path gets prepended to any file that you try and deal with, so something like /etc/mycnf/foo.cnf becomes /users/jeffcof/dev/bld-mysql/mysql/test/code/.../etc/mycnf/foo.cnf). This way, you have just one or two lines of changes; otherwise, we're reading in real-live config files from real-live platforms (as well as your specific tests that you may include),

3. Have your test dependency object allow "flags". So, for example, the production code may detect a loop and call "LoopDectected" in the dependency object. The production version of that may throw an exception, where the test version of that may just set a flag in the dependency object and return. Then, in the test itself, you can be sure the flag was set (i.e. a loop was detected).

In particular, I think #2 would make the tests a lot shorter and easier to extend. You just point a directory for a test, load it, and then look at the resulting object.

I very much like that you're validating specific authentications for specific files. Keep that, that's just what I would have done, and insures that the state you expect is just as anticipated after a set of configuration files are read.

Let me know what you think of this. Again, I do NOT want to have a "not invented here" syndrome, so if you really feel like this approach is better, please tell me why and let's chat.

In the case of GetLinuxOS, we wrote it initially, and then each time we changed the script due to a new platform, I added the actual config files for that platform. Makes sure that the new platform works, all older platforms work, and that it's easily extendable for new stuff down the road.

kapilash replied:
FWIW, I too did not like the tests. But I do not like the alternative either :cry: 

 I used an in-memory config file for the following reasons (in approximately decreasing order of priority):

1.  LICENSE and Copyright  -
 I do not know the ownership/copyright status of files like my.cnf from ubuntu/debian distributions of the package. Can I just copy it into our codebase (with minor modifications)?

2. Even when I am using real files and real directories, I will still need some sort of mapping between the test files and test directories and the entries in the file.
The tests are typically run from the folder $HOME/dev/bld-mysql/mysql/intermediate/Arch.
So, the test should introduce a mapping like one of the following

      /etc/mysql/mysql.d/mysqld.conf     ---> <TEST-DATA-DIR>/etc/mysql/mysql.d/mysqld.cnf 
    
     No matter how I compute the above mapping at the time of test execution, I will impose some restrictions on the test execution.  (testrunner must be run only from the directory that is at a fixed location relative to the test data - or copy the entire directory with test-data into the current folder).

3. To understand what a test does, I need to look at multiple files - test code, test data, each of the included files (or document it and hope that the documentation, test code and the test data are all in sync). 

4. A chance of false-positives is higher in case of test-data in files. (Test reports success but actually it is not testing the required data). This is again because the test logic is determined by multiple artifacts. 

I do think mocking is the right way. But I understand some of the above are my own personal preferences.  I just wanted to present the reasons I chose. Please let me know if you prefer the files.

jeffaco replied:
Addressing your comments:

1. These are open source files. I don't think you need to worry about copyright. Besides, the copyright really belongs to MySQL, and MySQL has an MIT license (very lax). Furthermore, we're not "redistributing" as part of our kit, we're just using them for testing. I don't think #1 is compelling. Furthermore, by not checking actual files from distributions, how do we KNOW we're handling them properly?

2. No, that's not the case. Look at how GetLinuxOS deals with it. The Makefile figures out the "base pwd" for our sources, then builds a directory path to them. This, ultimately, is passed down to GetLinuxOS.sh, which uses those vs. the actual system files on the current system. You can do something virtually identical to that in this case, I think (except as a "prefix" since some files may refer to other files by path).

3. Yes, this is true. But this is true anyway; to verify that data files are processed properly, you need to look at them and see what instances they set up. This isn't a reason to not have robust tests with real-life examples.

4. I don't understand this, sorry. You read the files, then you check the resulting auth object for correctness. It's either correct or it isn't correct. If it isn't correct, then you need to figure out why. To me, there's no better way to handle this than with real-life examples. In this way, we KNOW that actual examples that we include truly work (just like we do with GetLinuxOS.sh). Imagine if GetLinuxOS.sh didn't use real life examples - how would we know it was doing the right thing? I think the same is true in this case.

---
In reply to: [66900085](https://github.com/Microsoft/MySQL-Provider/pull/5#discussion_r66900085) [](ancestors = 66900085)

jeffaco commented:
I'd like to see another iteration please with these comments address, thanks.

jeffaco commented:
:clock1:

jeffaco commented on source/code/providers/support/mycnf-scanner.cpp:
```diff
@@ -0,0 +1,244 @@
+        {
+            std::wstringstream ss;
+            ss << L"Encountered " << e.What() << L" when trying to read entried from " << directory.Get();
+            SCX_LOGERROR(m_log, ss.str());
```
You're logging an error and not aborting. Thus, processing will continue even though we can't process properly. We should probably just stop, leaving the existing auth file alone.

kapilash replied:
As I mentioned above,  I was deliberately doing that.  MySQL documentation does not say what it will do if ,say, the directory does not present. If it ignores such errors, we will be wrong in throwing exception and bailing out. In general, we must be more lenient in reading the cnf than MySQL. 
Anyways, I will change it.  The problem here is the options file is under-specified. I will send out a new iteration.

kapilash replied:
edited out some atrocious grammar.

kapilash commented:
I've sent out a new iteration. PTAL.

The following are the main changes since the earlier version:

1. MySQL_ConfigurationScanner class holds state.
2. Test Code uses real files and directories instead of in-memory files.
          (cnf files for rhel/sles/debian are used. In addition to other test cnf files)
3. Scanner now throws Exceptions instead of ignoring missing files or recursive loops.

As a result of 3, the following are the ways in which Scanner differs from the mysql behavior:

 **path does not exist** 
MySQL ignores missing files but raises error for missing directories.
 That is, if the file referred to in an include directive does not exist, MySQL ignores it. (There is not even a warning). The same is not true in case of ```include-dir``` directive. If the directory referered to is missing or is not accessible, mysql throws and the server does not start. 

Scanner fails for both cases.

 **handling recursion**
MySQL identifies loops based on [```max_recursion_level(10)```](https://github.com/mysql/mysql-server/blob/5.7/mysys_ssl/my_default.cc#L874).  Once the loop count crosses [```max_recursion_level```](https://github.com/mysql/mysql-server/blob/5.7/mysys_ssl/my_default.cc#L925) it comes out of the loop without processing any further. 

Scanner keeps a set of files visited and throws an exception when it encounters the same file again.


Apart from looking at the MySQL code, the above behavior of MySQL was validated on ubuntu 16.

1. ```/etc/init.d/mysqld stop```
2. Change cnf 
3. ```/etc/init.d/mysqld start```
(Check ```systemctl status mysql.service``` to see errors/warnings if any)

jeffaco commented on build/Makefile:
```diff
@@ -501,7 +505,11 @@ testrun : test
-test : TEST_STATUS $(SCXPAL_INTERMEDIATE_DIR) $(INTERMEDIATE_DIR)/testrunner
+TEST_DATA_PREP:
+	@echo "========================= Performing Test data preparation"
+	cp -R $(TEST_DIR)/omi_preexec/testdata/cnfs $(INTERMEDIATE_DIR)/.
```
Is this needed? The files aren't being changed, are they? If not, could you just use them in place without copying them?

kapilash replied:
you mean via a soft link instead of a copy, correct?

Just to be on same page, the paths in the includedirs and include in all the files will be affected.

jeffaco replied:
Yeah, that's what I thought. I think that, in the test, you should pass the prefix that should be added to all "open" files. So, for example, set up the tests like:

test/code/omi_preexec/testdata/cnfs/rhel/etc/mysql/mysql.cnf

You would pass in the full path to test/code/omi_preexec/testdata/cnfs/rhel as a prefix in the test object. The test object would prepend that to any !include or !includedir. In this way, you could check in REAL (unmodified) files from vendors and have them work perfectly, assuming your code is correct. I'm not wild changing the vendor files to work under test unless we have to and, in this case, we can make them work properly.

---
In reply to: [67187563](https://github.com/Microsoft/MySQL-Provider/pull/5#discussion_r67187563) [](ancestors = 67187563)

kapilash replied:
I think I am missing something here.

There is only ONE vendor file that uses include and includedir. That is debian(ubuntu). No body else uses those directives. 

Rest of the cnf files are handwritten to test particular features. 
I am missing something here. 
Firstly, I do not see why the vendor files are to be treated as some sort of gospel - users of MySQL can and do change these.
Secondly, include and includedir are a relatively rare uses.

Also, please see my other comment about how we already differ from how MySQL treats these values and we do.

jeffaco replied:
The idea here is that we want to make sure the vendor files, as shipped, work properly. Currently, they do not, and we'd like to make sure that they do, preferably without modifications.

Also, I would expect that, over time as MySQL itself is modified and more versions come out, we'll check in more vendor files (and associated unit tests). That's exactly what happened with GetLinuxOS, anyway.

---
In reply to: [67194969](https://github.com/Microsoft/MySQL-Provider/pull/5#discussion_r67194969) [](ancestors = 67194969)

jeffaco commented on source/code/omi_preexec/mycnf-scanner.cpp:
```diff
@@ -0,0 +1,251 @@
+        if (!ins.second)
+        {
+            std::wstringstream ss;
+            ss << file.Get() << L"is used in !include more than once.";
```
Need a space after ", like: L" is used in ..."? Without a space, would we see something like:   file.cnfis used ...?

jeffaco commented on source/code/omi_preexec/mycnf-scanner.cpp:
```diff
@@ -0,0 +1,251 @@
+        {
+            std::wstringstream ss;
+            ss << file.Get() << L"is used in !include more than once.";
+            throw MySQL_ConfigScannerException(ss.str(), SCXSRCLOCATION);
```
Using generic exceptions like this is a double edged sword:

Pro: It's easy, one exception serves all,
Con: Impossible to easily catch any specific exception for special handling

Just FYI. I presume you don't need to catch specific exceptions, so you probably don't care.

jeffaco commented on source/code/omi_preexec/mycnf-scanner.cpp:
```diff
@@ -0,0 +1,251 @@
+        const SCXCoreLib::SCXFilePath& directory, 
+        MySQL_Authentication& auth)
+    {
+        SCX_LOGHYSTERICAL(m_log, SCXCoreLib::StrAppend(L"LoadCnfFilesFromDirectory", directory.Get()));
```
Slightly better output, like: L"LoadCnfFilesFromDirectory loading: ", directory.Get() ... ?

jeffaco commented on source/code/omi_preexec/mycnf-scanner.cpp:
```diff
@@ -0,0 +1,251 @@
+            else
+            {
+                SCX_LOGHYSTERICAL(m_log,SCXCoreLib::StrAppend(iter->Get(), L" is not a cnf file"));
+            }
```
This seems needlessly restrictive. Does MySQL itself allow config files to be named anything (like MySQL.config?)  Is this a MySQL restriction, or are you imposing this yourself?

Also, what happens if an include directory contains other directories (even like "secondinclude.cnf")? It would pass your test. You probably shouldn't call LoadCnfFile on a directory, or anything that isn't a regular file for that matter.

kapilash replied:
The file extension is MySQL restriction. From the [documentation](http://dev.mysql.com/doc/refman/5.7/en/option-files.html), 

> Any files to be found and included using the !includedir directive on Unix operating systems must have file names ending in .cnf

And the word "must" is actually emphasized on that site.

> what happens if an include directory contains other directories (even like "secondinclude.cnf")? 

A directory with extension ".cnf" in the name? I did not check it. I do not know what MySQL does in such cases either.

jeffaco replied:
Easy enough to check if it's a directory and, at the least, not try to open it. Might be worth checking what MySQL would do in this case, so you can be consistent. They can either error out, or they can go read that directory for .cnf files.

---
In reply to: [67188834](https://github.com/Microsoft/MySQL-Provider/pull/5#discussion_r67188834) [](ancestors = 67188834)

kapilash replied:
OK. I can add the check. 
I know for sure that MySQL would not treat it like a directory. 

As I explained in earlier comment, MySQL is very very lenient as far as "!include"  is concerned.
It ignores a file that is not present (No error no warning). It ignores a non-readable file etc.

jeffaco commented on source/code/omi_preexec/mycnf-scanner.cpp:
```diff
@@ -0,0 +1,251 @@
+        m_ports.clear();
+        LoadCnfFile(cnfFile, auth);
+
+        std::vector<unsigned int> authPorts;
```
These comments came from original code, and it's worth including here (took me a minute to understand what was happening here):

// We've added what must be added, but we need to delete unreferenced ports
// (Loop through authentication ports and remove anything we didn't read from configuration

jeffaco commented on test/code/omi_preexec/config_scanner_test.cpp:
```diff
@@ -0,0 +1,231 @@
+        portBindingMap[3306] = "127.0.0.1";
+
+        MySQL::MySQL_ConfigurationScanner scanner;
+        scanner.Scan(L"cnfs/rhel/my.cnf", auth);
```
Oh, I see how this works, and why you're copying the files from their source location to the intermediate directory.  But what happens if my.cnf includes something else? Do you have to modify the checked in config files to be careful about paths?

If so, I would have preferred that you pass in a directory prefix to the test dependency object and have that prefix any directory (from !include or !includedir), thus allowing you to run "in place", without copying files around. I haven't looked at the actual configuration files yet, so I don't know if you modified this from what was shipped.

jeffaco replied:
I really liked how this worked out, by the way. The tests are MUCH smaller and simpler this way, and it's much easier to add new tests. Very nicely done.

---
In reply to: [67190807](https://github.com/Microsoft/MySQL-Provider/pull/5#discussion_r67190807) [](ancestors = 67190807)

jeffaco commented:
> 

I'm not wild that you had to edit these. I think it would be nicer if the files were EXACTLY from the source system, and the test prepended the location. Then you copy the files from the platform into a directory, create the test, and you're done. You don't need to modify the actual vendor files to make them work.

Also, your directory naming can be improved. In this case, you have test/code/omi_preexec/testdata/cnfs/debian.

What version of Debian? 5? 6? 7? Really Debian, or Ubuntu (I suspect this is from Ubuntu)?

Suggest you use naming conventions like what I used in GetLinuxOS for directories, like Unbutu_16.04, etc, so they are very clear. Similar for rhel and sles directories (i.e. rhel_7.0, etc).

You may want to consider naming conventions that clarify vendor files from internal test files, too. But that's a nit, and is up to you. To clarify, perhaps something like:

test/code/omi_preexec/testdata/cnfs/vendor/rhel_7.0/...
test/code/omi_preexec/testdata/cnfs/derived/testname/...

But that's less important than at least naming the vendor files to at least include the version.

---
Refers to: test/code/omi_preexec/testdata/cnfs/debian/my.cnf:22 in dfed34b. [](commit_id = dfed34b701fe10e612efcc18c67e8fffb50e160a, deletion_comment = False)

jeffaco commented:
Okay, I'm going to sign off. I'll trust that you'll incorporate my comments appropriately. Unless you want me to look at followup changes, go ahead and merge to mainline.

However, please notify me when you commit, as I'll need to do a build and get this included in the current SCXcore kits (to properly support Ubuntu 16 going forward).

Thanks, this was nicely done.

kapilash commented:
1. Changed the log statements
2. Added check for skipping if the entry is a directory (when iterating in a directory)
3. Moved test files to vendor/derived
4. Added version info in the directory containing test cnfs (rhel_7.2 instead of rhel etc)
5. Test hook takes prefix appends it to the file when needed, before opening. (This allows us to avoid changing the cnf files)

I've fixed everything. Ran tests and Merging this.
